### DISCUSSION TITLE
STM32F7  DMA + ADC question

### DISCUSSION THREAD

**Discussion by ddkn** [zephyrproject-rtos/zephyr#33221] [Q&A]
Hello,

I am a bit unfamiliar with the ZephyrOS, but I was sifting through the documentation about a project I am working on for an experiment in grad school reading a single channel off of the ADC in asynchronous mode (possibly with DMA). The reason for the RTOS convenience of a microSD.

I have a Nucleo STM32F767si and I am trying to read data that could be about 20kHz, where the timing is important. I am unsure if the STM32F7 has DMA support or how to set it up within the confines of the Zephyr, sifting through some of the issues I see there was some support added for the F4?. Ideally, I would like to store data into a large buffer or two (or Sequence? if that is how the terminology works here?), and store that to an SD card later. However, I am unsure based on the documentation, if the api is very limited or I am misreading something.

In short:
* ADC driver to store into buffer at some acquisition rate asynchronously or by some hw timer? I don't think the sample example for drivers/adc shows this. 
* Is the stm32f7 DMA suppoorted

Is it just better to implement these things manually outside of zephyr api's and just use the zephyr api for microSD? Or use something like NuttX? (again, not my expertise either)

Is it possible to have some guidance? Or direction? I would really appreciate it.

Cheers,

**Comment by ddkn** [comment]
I found a nice tutorial to follow through. Avoiding the Zephyr DMA drivers at the moment (in part to understand them better). In a memory-to-memory transfer on a basic uint16_t array, I can generate successful transfer complete interrupts, but no data is ever transferred from the source to the destination. 

According to the STM32 datasheet and setting up to use DMA2, Channel 0 on Stream 0 it says I should be using `DMA2_Stream0->PAR` and `DMA2_Stream0->M0AR`. Using the appropriate data size, and amount of data to transfer on both memory locations for the the DMA 

Has anyone run into a situation where the memory from the DMA is triggering transfer complete interrupts, but no data is actually transferred? I double checked my settings, and am wondering if there is something in zephyr preventing or blocking those transfers? Some people mention disabling the FIFO or D-Cache or I-Cache, but those seemed to be not enabled in the Kconfig anyway.

Any help or direction would be appreciated. I will keep this thread updated.
